QAT v. 0.1 (#1230)
microsoft/qsharp-compiler: src/Passes/docs/src/DiscussionDocs/ThoughtsOnQubitManagement.md

## Possible lifetime management of qubits

We can think of a few different life-time management scenarios:

1. Scoped lifetime management
2. Lifetime limited to that of a child scope
3. Lifetime limited to the lifetime of a containing object
4. Fully reference counted.

So far we have only considered 1 and 4. Using allocators and deallocators (in a languge that support move semantics) allows us to implement all four scenarios.

Forcing reference counting will not allow us to do 2 and only to some degree 3. Further, we would only be able to map back to 1 if the count adds up at compile time. While 4 provides the most secure scenario there are many cases where 1-3 are desirable at least in ordinary resource management scenarios.

## Qubit allocation in Q#

Currently Q# implements 1 and it is appropriate to briefly consider how it currently uses `aloocate` and `release` to do so: Acquiring qubit access in Q# is done through a `use` statement which assigns the qubit to the current scope:

```

>     operation Main(): Unit
>     {

Reviewer: ```suggestion
    operation RunMain() : Unit {
```

Author: This document is unfit to go into main. Deleted.

        use q1 = Qubit();

        /// ... The underlying qubit ID is reserved through

>         /// the body of this function ...

Reviewer: `RunMain` is declared as an operation, not as a function.
```suggestion
        /// the body of this operation...
```

        /// ... and released right before the exit of the scope.
    }
```

The use statement creates a allocate instruction and release instructions are inserted at the exit of the scope. However, one issue that is encountered in the Q# language is that it becomes hard for the frontend to provide garantuees that the qubit does not escape the scope. A good example of a hard case is

```

>   operation Test() : Unit
>   {

Reviewer: ```suggestion
  operation Test() : Unit {
```

Author: This document is unfit to go into main. Deleted.

    use (q1, q2) = (Qubit(), Qubit());
    mutable arr = [q1, q2];

>     for(i in 0..10)
>     {

Reviewer: ```suggestion
    for i in 0..10 {
```

Author: This document is unfit to go into main. Deleted.

      use n = Qubit();

>       arr /w= 0 <- q[1];
>       arr /w= 1 <- n;

Reviewer: ```suggestion
      arr w/= 0 <- q[1];
      arr w/= 1 <- n;
```

Author: This document is unfit to go into main. Deleted.

    }

    H(arr[0]);

  }
```

> This case makes it hard to analyse which qubits stays in scope and which are escaping. We note that we have two scopes where qubits can escape: The Test function and the for loop.

Reviewer: Test is not declared as a function, but as an operation.

```suggestion
This case makes it hard to analyze which qubits stays in scope and which are escaping. We note that we have two scopes where qubits can escape: The `Test` operation and the `for` loop.
```

For the misspelling, adding a cspell CI check is highly suggested; the Q# samples repo has an example of how to do so via GH Actions.
